src/lib.rs

```diff
@@ -299,7 +299,7 @@ impl Client {
     }
 
     /// Starts a new SELECT/DDL query.
-    pub fn query(&self, query: &str) -> query::Query {
+    pub fn query<'a>(&self, query: &'a str) -> query::Query<'a> {
```

Comment:
Avoiding allocations here is not worth introducing such a breaking change in public API.

* Allocations are (usually) done in the same thread as deallocations, so for most modern allocators it's cheap (any performance-focused code anyway will use mimalloc/tcmalloc/jemalloc)
* Allocations here are rare operations compared to decoding incoming rows
* Constructing a query is negligible compared to HTTP and CH

`Query`/`Watch` can be stored and even sent; I'm not sure we should prevent a such usage.